nucypher/blockchain/eth/actors.py

```diff
@@ -1493,76 +1493,75 @@ def was_this_evidence_evaluated(self, evidence) -> dict:
         return receipt
 
 
-class StakeHolder(Staker):
+class Wallet:
+    """
+    Account management abstraction on top of blockchain providers and external signers
+    """
+    class UnknownAccount(KeyError):
+        pass
 
-    banner = STAKEHOLDER_BANNER
+    def __init__(self,
+                 client_addresses: set = None,
+                 provider_uri: str = None,
+                 signer=None):
 
-    class StakingWallet:
+        self.__client_accounts = list()
+        self.__transacting_powers = dict()
 
-        class UnknownAccount(KeyError):
-            pass
+        # Blockchain
+        self.blockchain = BlockchainInterfaceFactory.get_interface(provider_uri)
+        self.__signer = signer
+
+        self.__get_accounts()
+        if client_addresses:
+            self.__client_accounts.extend([a for a in client_addresses if a not in self.__client_accounts])
+
+    @validate_checksum_address
+    def __contains__(self, checksum_address: str) -> bool:
+        return bool(checksum_address in self.accounts)
 
-        def __init__(self,
-                     registry: BaseContractRegistry,
-                     client_addresses: set = None,
-                     signer=None):
+    @property
+    def active_account(self) -> str:
+        return self.blockchain.transacting_power.account
 
-            self.__local_accounts = dict()
-            self.__client_accounts = set()  # Note: Account index is meaningless here
-            self.__transacting_powers = dict()
+    def __get_accounts(self) -> None:
+        if self.__signer:
+            signer_accounts = self.__signer.accounts
+            self.__client_accounts.extend([a for a in signer_accounts if a not in self.__client_accounts])
```

Comment:
This filtration feels a bit clunky to me - is it possible to use a dict/set to track unique addressable accounts here?

Reply:
It's clunky, but sets are definitely not an option here. Sets have an undefined order, which changes with every execution. See this commit of this very same PR (b18b9d3a2d843751280cd24f49f4b63057d259bd). It took me a lot to debug the problems caused by sets.

Reply:
Despite your aversion to sets - How about a dictionary?

Reply:
It's not aversion – I love sets; they're simply not the right collection here as we want to preserve order of insertion. Yes, we can use dicts, I thought about that too, but to be honest, I also find that weird: we'd be using a strange feature of dicts which only exists from python 3.7 and we wouldn't be using dict values. Current solution is clunky but it's declaratively precise.

Reply:
Nothing strange about it - It's actually considered an implementation detail as of python 3.6.0 (and earlier), but can be completely considered to be ordered for implementation as of python 3.7.

See this StackOverflow Question: <https://stackoverflow.com/questions/39980323/are-dictionaries-ordered-in-python-3-6>

Reply:
Aside from the dictionary implementation, IMO if you need a unique insertion-ordered collection and are forced to "manually" use filtration methods, this may indicate code inefficiency and make for some flaky features.

Reply:
Let's add a TODO here. I didn't use a dictionary on purpose since I didn't find it elegant at that time, given that there was no value to store. However, I had an issue in my head with the behavior of Wallet, and perhaps in this case the values of the dict could be of use. Basically, we need to discern between accounts from the signer and accounts from a non-signing provider (e.g. Infura), as the latter won't be able to transact. This will simply improve error-handling, so it's not urgent.

Reply:
:+1: Sounds good!